### PR TITLE
Bug28223 diagnostic 035

### PR DESCRIPTION
**Pull request by nmathewson**

### CODE REVIEW

File: src/feature/nodelist/microdesc.c [hunk 1]
```diff
@@ -221,6 +221,13 @@ dump_microdescriptor(int fd, microdesc_t *md, size_t *annotation_len_out)
   }
 
   md->off = tor_fd_getpos(fd);
+  const char *nulpos = memchr(md->body, 0, md->bodylen);
+  if (BUG(nulpos)) {
+    log_warn(LD_BUG, "About to dump a NUL into a microdescriptor file. "
+             "offset %"PRId64", bodylen %zu, nul position %zu",
+             (int64_t)md->off, md->bodylen,
+             (size_t)(nulpos - md->body));
```
Comment (teor2345) [inline src/feature/nodelist/microdesc.c]: Can we log the escaped() context of the NUL?
Maybe 10 bytes either side of the NUL?

Comment (teor2345) [reply src/feature/nodelist/microdesc.c]: Also, can we log some identifier from the microdescriptor?

Comment (teor2345) [reply src/feature/nodelist/microdesc.c]: Should we call warn_if_nul_found() here?

File: src/feature/nodelist/microdesc.c [hunk 2]
```diff
@@ -221,6 +221,13 @@ dump_microdescriptor(int fd, microdesc_t *md, size_t *annotation_len_out)
   }
 
   md->off = tor_fd_getpos(fd);
+  const char *nulpos = memchr(md->body, 0, md->bodylen);
+  if (BUG(nulpos)) {
+    log_warn(LD_BUG, "About to dump a NUL into a microdescriptor file. "
+             "offset %"PRId64", bodylen %zu, nul position %zu",
```
Comment (teor2345) [inline src/feature/nodelist/microdesc.c]: Windows says:
"error: unknown conversion type character 'z' in format"
https://ci.appveyor.com/project/torproject/tor/builds/23576412/job/80p2k2cumplatqx9#L1468

File: src/feature/nodelist/microdesc.c [hunk 3]
```diff
@@ -480,6 +487,17 @@ microdesc_cache_clear(microdesc_cache_t *cache)
   cache->bytes_dropped = 0;
 }
 
+static void
+warn_if_nul_found(const char *inp, size_t len, const char *description)
+{
+  const char *nul_found = memchr(inp, 0, len);
+  if (BUG(nul_found)) {
+    log_warn(LD_BUG, "Found unexpected NUL while reading %s, at "
+             "position %zu/%zu.",
```
Comment (teor2345) [inline src/feature/nodelist/microdesc.c]: Windows probably doesn't like this z, either.

File: src/feature/nodelist/microdesc.c (hunk 1 above)
Comment (nmathewson) [reply src/feature/nodelist/microdesc.c]: I'll try to do the other stuff; at this stage, there is no identifier from the md that we could log.

**Pull request merged**